src/transformers/modeling_utils.py

```diff
@@ -1569,6 +1581,82 @@ def retrieve_modules_from_names(self, names, add_prefix=False, remove_prefix=Fal
 
         return retrieved_modules
 
+    @classmethod
+    def _load_state_dict_into_model_low_mem_p1(cls, state_dict):
+        """
+        This is an experimental 2-part function that loads the model using ~1.x model size CPU memory
+
+        Part 1:
+
+        1. save which state_dict keys we have
+        2. drop state_dict
+        """
+        require_version_core("torch>=1.9")
+        if is_deepspeed_zero3_enabled():
+            raise ValueError("low_cpu_mem_usage arg cannot be used with DeepSpeed ZeRO-3")
+
+        # save the keys
+        loaded_state_dict_keys = [k for k in state_dict.keys()]
+        del state_dict  # free CPU memory - will reload again later
+        return loaded_state_dict_keys
+
+    @classmethod
+    def _load_state_dict_into_model_low_mem_p2(cls, model, loaded_state_dict_keys, resolved_archive_file):
+        """
+        This is an experimental 2-part function that loads the model using ~1.x model size CPU memory
+
+        Part 2:
+
+        3. switch to the meta device all params/buffers that are going to be replaced from the loaded state_dict
+        4. load state_dict 2nd time
+        5. replace the params/buffers from the state_dict
+
+        Currently, it doesn't handle missing_keys, unexpected_keys, mismatched_keys. It can't handle deepspeed.
+        """
+
+        # a helper util to find the last sub-module and the param/buffer name
+        def find_submodule_and_param_name(model, long_key):
+            split_key = long_key.split(".")
+            submodule = model
+            while len(split_key) > 1:
+                if hasattr(submodule, split_key[0]):
+                    submodule = getattr(submodule, split_key[0])
+                    del split_key[0]
+                else:
+                    submodule = None
+                    break
+            return submodule, split_key[0]
+
+        # dematerialize param storage for keys that are going to be replaced by state_dict, by
+        # putting those on the meta device
+        for k in loaded_state_dict_keys:
+            submodule, param_name = find_submodule_and_param_name(model, k)
+            if submodule is not None:
+                # selectively switch to the meta device only those params/buffers that will
+                # be next replaced from state_dict. This a complex way to do p.to_("meta")
+                # since we have no in-place to_ for tensors.
+                new_val = getattr(submodule, param_name)
+                if isinstance(new_val, torch.nn.Parameter):
+                    # isinstance returns False for Params on meta device, so switch after the check
+                    new_val = torch.nn.Parameter(new_val.to("meta"))
+                else:
+                    new_val = new_val.to("meta")
```

Comment:
won't that approach be a problem for buffers? Would they now become unregistered. If you remember my initial attempt was manually re-registering the buffer and you suggested not to do that, but to update the existing buffer instead.

On the other hand given that 1.10 is imminent, it's probably OK to require pt-1.9. Perhaps we can start with 1.9, and if a lot of users request to support lower versions then we can switch to your proposal?

Reply:
Ho right you would need to re-register the buffers... So doing `setattr(submodule, param_name, None)` then haha

But I agree that this is most likely not worth the problem. Just wanted to mention it in case.